File: 1469-find-all-the-lonely-nodes/1469.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)
        self.dfs(root.right)

        if root.left and not root.right:
            self.ans.append(root.left.val)
        elif root.right and not root.left:
            self.ans.append(root.right.val)

    def getLonelyNodes(self, root: Optional[TreeNode]) -> List[int]:
        self.ans = []
        self.dfs(root)
        return self.ans

File: 1469-find-all-the-lonely-nodes/1469.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<int> ans;
    void dfs(TreeNode* node) {
        if (!node)
            return;
        if (node->right && !node->left) {
            ans.push_back(node->right->val);
        } else if (!node->right && node->left) {
            ans.push_back(node->left->val);
        }
        dfs(node->left);
        dfs(node->right);
    }
public:
    vector<int> getLonelyNodes(TreeNode* root) {
        dfs(root);
        return ans;
    }
};
